### PR TITLE
/.well-known/core discovery implementation

### PR DESCRIPTION
**Pull request by WAvdBeek**

### CODE REVIEW

File: security/oc_acl.c [hunk 1]
```diff
@@ -365,8 +365,12 @@ oc_sec_check_acl(oc_method_t method, oc_resource_t *resource,
...
       granted.
   */
-  if (pstat->s == OC_DOS_RFOTM && method == OC_GET &&
-      ((oc_string_len(resource->uri) == 8 &&
+  if (pstat->s == OC_DOS_RFOTM && method == OC_GET && (
+#ifdef OC_WKCORE
+       (oc_string_len(resource->uri) == 17 &&
+        memcmp(oc_string(resource->uri), "/.well-known/core", 17) == 0) ||
+#endif
+       (oc_string_len(resource->uri) == 8 &&
```
Comment (tiniuclx) [inline security/oc_acl.c:373]: Would you mind updating the comment just above?

**Review by tiniuclx — CHANGES_REQUESTED** [review]
Changes look good overall, but there is a potential memory corruption bug I would like to see addressed.

File: api/oc_discovery.c [hunk 2]
```diff
@@ -749,9 +757,134 @@ oc_core_discovery_handler(oc_request_t *request, oc_interface_mask_t iface_mask,
...
+  size_t response_length = 0;
+  int matches = 0;
+
+  /* enable this later*/
+  /*
+  if (request->origin && request->origin->version != APPLICATION_LINK_FORMAT) {
+    request->response->response_buffer->code =
+      oc_status_code(OC_STATUS_BAD_REQUEST);
+    return;
+  } */
```
[the remark this answers is not in the document]
Comment (WAvdBeek) [reply api/oc_discovery.c:776]: no, i will enable this

File: api/oc_discovery.c [hunk 3]
```diff
@@ -749,9 +757,134 @@ oc_core_discovery_handler(oc_request_t *request, oc_interface_mask_t iface_mask,
...
+  for (i = 0; i < (int)oc_string_array_get_allocated_size(resource->types);
+    i++) {
+    size_t size = oc_string_array_get_item_size(resource->types, i);
+    const char* t =
+      (const char*)oc_string_array_get_item(resource->types, i);
+    if (strncmp(t, "oic.d", 5) == 0)
+    {
+      /* take the first oic.d.xxx in the oic/d of the list of resource/device types */
+      strncpy(rt_device, t, size);
+      rt_device[size] = '\0';
```
[the remark this answers is not in the document]
Comment (WAvdBeek) [reply api/oc_discovery.c:813]: rewrote the code, it pulls now the data directly from the header. no copying of data anymore.

File: security/oc_acl.c (hunk 1 above)
Comment (WAvdBeek) [reply security/oc_acl.c:373]: done

File: api/oc_discovery.c [hunk 4]
```diff
@@ -767,34 +771,31 @@ oc_wkcore_discovery_handler(oc_request_t* request, oc_interface_mask_t iface_mas
...
     if (strncmp(key, "rt", key_len) == 0) {
-      strncpy(rt_request, value, value_len);
-      rt_request[value_len] = '\0';
+      rt_request = value;
       rt_len = (int)value_len;
     }
   }
 
-  if (strncmp(rt_request, "oic.wk.res", rt_len) == 0) {
+  if (rt_request != 0 && strncmp(rt_request, "oic.wk.res", rt_len) == 0) {
```
Comment (tiniuclx) [inline api/oc_discovery.c]: I like this a lot more!

**Review by tiniuclx — APPROVED** [review]

**Pull request merged**